### ISSUE TITLE
Can wrongly assign object with incomplete mapped type to interface with mapped type member

### ISSUE THREAD

**Issue by meustrus** [microsoft/TypeScript#29190]
**TypeScript Version:**  3.2.2, Playground

**Search Terms:** mapped type assign

**Code**

```ts
type MappedType<KeyUnion extends number> = {
    mapped: { [key in KeyUnion]: string }
}
interface MappedInterface<KeyUnion extends number> {
    mapped: { [key in KeyUnion]: string }
}

// expected: no error; actual: no error
let onlyOne = {
    mapped: { 1: "one" }
}
let oneOrTwo = {
  mapped: { 1: "one", 2: "two" }
}
let onlyOneWithType: MappedType<1> = onlyOne
let onlyOneWithInterface: MappedInterface<1> = onlyOne
let oneOrTwoWithType: MappedType<1 | 2> = oneOrTwo
let oneOrTwoWithInterface: MappedInterface<1 | 2> = oneOrTwo

// expected: error; actual: error
let oneOrTwoWithTypeFromOnlyOne: MappedType<1 | 2> = onlyOne
let oneOrTwoWithTypeFromOnlyOneWithType: MappedType<1 | 2> = onlyOneWithType
let oneOrTwoWithTypeFromOnlyOneWithInterface: MappedType<1 | 2> = onlyOneWithInterface
let oneOrTwoWithInterfaceFromOnlyOne: MappedInterface<1 | 2> = onlyOne
let oneOrTwoWithInterfaceFromOnlyOneWithType: MappedInterface<1 | 2> = onlyOneWithType

// expected: error; actual: no error
let oneOrTwoWithInterfaceFromOnlyOneInterface: MappedInterface<1 | 2> = onlyOneWithInterface
```

**Expected behavior:**
Compile error along the lines of:

```
Type 'MappedInterface<1>' is not assignable to type 'MappedInterface<1 | 2>'.
  Types of property 'mapped' are incompatible.
    Property '2' is missing in type '{ 1: string; }' but required in type '{ 1: string; 2: string; }'.
```

**Actual behavior:**
Assignment to `oneOrTwoWithInterfaceFromOnlyOneInterface` compiles.

[**Playground Link**](http://www.typescriptlang.org/play/#src=type%20MappedType%3CKeyUnion%20extends%20number%3E%20%3D%20%7B%0D%0A%20%20%20%20mapped%3A%20%7B%20%5Bkey%20in%20KeyUnion%5D%3A%20string%20%7D%0D%0A%7D%0D%0Ainterface%20MappedInterface%3CKeyUnion%20extends%20number%3E%20%7B%0D%0A%20%20%20%20mapped%3A%20%7B%20%5Bkey%20in%20KeyUnion%5D%3A%20string%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20expected%3A%20no%20error%3B%20actual%3A%20no%20error%0D%0Alet%20onlyOne%20%3D%20%7B%0D%0A%20%20%20%20mapped%3A%20%7B%201%3A%20%22one%22%20%7D%0D%0A%7D%0D%0Alet%20oneOrTwo%20%3D%20%7B%0D%0A%20%20mapped%3A%20%7B%201%3A%20%22one%22%2C%202%3A%20%22two%22%20%7D%0D%0A%7D%0D%0Alet%20onlyOneWithType%3A%20MappedType%3C1%3E%20%3D%20onlyOne%0D%0Alet%20onlyOneWithInterface%3A%20MappedInterface%3C1%3E%20%3D%20onlyOne%0D%0Alet%20oneOrTwoWithType%3A%20MappedType%3C1%20%7C%202%3E%20%3D%20oneOrTwo%0D%0Alet%20oneOrTwoWithInterface%3A%20MappedInterface%3C1%20%7C%202%3E%20%3D%20oneOrTwo%0D%0A%0D%0A%2F%2F%20expected%3A%20error%3B%20actual%3A%20error%0D%0Alet%20oneOrTwoWithTypeFromOnlyOne%3A%20MappedType%3C1%20%7C%202%3E%20%3D%20onlyOne%0D%0Alet%20oneOrTwoWithTypeFromOnlyOneWithType%3A%20MappedType%3C1%20%7C%202%3E%20%3D%20onlyOneWithType%0D%0Alet%20oneOrTwoWithTypeFromOnlyOneWithInterface%3A%20MappedType%3C1%20%7C%202%3E%20%3D%20onlyOneWithInterface%0D%0Alet%20oneOrTwoWithInterfaceFromOnlyOne%3A%20MappedInterface%3C1%20%7C%202%3E%20%3D%20onlyOne%0D%0Alet%20oneOrTwoWithInterfaceFromOnlyOneWithType%3A%20MappedInterface%3C1%20%7C%202%3E%20%3D%20onlyOneWithType%0D%0A%0D%0A%2F%2F%20expected%3A%20error%3B%20actual%3A%20no%20error%0D%0Alet%20oneOrTwoWithInterfaceFromOnlyOneInterface%3A%20MappedInterface%3C1%20%7C%202%3E%20%3D%20onlyOneWithInterface)

**Related Issues:** possibly [13178: Issue in mapped types inference](https://github.com/Microsoft/TypeScript/issues/13178)

**Comment by weswigham** [comment]
Based on how this applies to interfaces but not type aliases, I'd guess this is a bug with interface type parameter variance probing.

Yeah. if you enable `strictFunctionTypes` you get the error you expect - the issue is variance probing is short circuiting to `covariant` outside of strict function types mode for interfaces, which is obviously incorrect - key positions (like this) are contravariant, even if functions aren't checked strictly.

**Comment by EvilSpirit39** [comment]
It can't reproduce in version 3.4.0-rc or after. 

Maybe already fixed?

**Comment by meustrus** [comment]
You're right, it looks like this has been fixed for some time now. Nice!
